spring-session/src/integration-test/groovy/org/springframework/session/data/couchbase/BasicSpec.groovy

```diff
@@ -0,0 +1,186 @@
+
+    @Shared
+    private RestTemplate rest = new RestTemplate()
+    @Autowired(required = false)
+    private CouchbaseTemplate couchbase
+    @Autowired
+    private EmbeddedWebApplicationContext context
+    private int extraInstancePort
+    private ApplicationInstance instance
+    // Cannot store cookie in thread local because some tests starts more than one app instance. CANNOT run tests in parallel.
+    private String currentSessionCookie
+
+    void setup() {
+        createBucketIndex()
+        clearBucket()
+    }
+
+    void cleanup() {
+        clearSessionCookie()
+        stopApplication()
+    }
+
+    protected void startApplication(Class<?> applicationClass) {
+        URL[] urls = [new File('/build/classes/test').toURI().toURL()]
+        def classLoader = new URLClassLoader(urls, getClass().classLoader)
+        def runnerClass = classLoader.loadClass(ApplicationInstanceRunner.class.name)
+        def runnerInstance = runnerClass.newInstance()
+        instance = new ApplicationInstance(runnerClass, runnerInstance)
+        runnerClass.getMethod('setTestApplicationClass', Class).invoke(runnerInstance, applicationClass)
+        runnerClass.getMethod('run').invoke(runnerInstance)
+        extraInstancePort = runnerClass.getMethod('getPort').invoke(runnerInstance) as int
+    }
+
+    protected void stopApplication() {
+        if (instance) {
+            instance.runnerClass.getMethod('stop').invoke(instance.runnerInstance)
+            instance = null
+        }
+    }
+
+    protected boolean currentSessionExists() {
+        return couchbase.exists(getCurrentSessionId())
+    }
+
+    protected void setSessionAttribute(Message attribute) {
+        post('session/attribute', attribute, getPort())
+    }
+
+    protected void deleteSessionAttribute() {
+        delete('session/attribute', getPort())
+    }
+
+    protected ResponseEntity<Message> getSessionAttribute() {
+        return get('session/attribute', Message, getPort())
+    }
+
+    protected void setSessionBean(Message attribute) {
+        post('session/bean', attribute)
+    }
+
+    protected ResponseEntity<Message> getSessionBean() {
+        return get('session/bean', Message)
+    }
+
+    protected void invalidateSession() {
+        delete('session')
+    }
+
+    protected String setPrincipalSessionAttribute() {
+        return post('session/principal', null, getPort(), String).body
+    }
+
+    protected String setPrincipalSessionAttributeToExtraInstance() {
+        return post('session/principal', null, extraInstancePort, String).body
+    }
+
+    protected ResponseEntity<List<String>> getPrincipalSessions() {
+        return get('session/principal', List, getPort())
+    }
+
+    protected void clearSessionCookie() {
+        currentSessionCookie = null
+    }
+
+    protected void clearBucket() {
+        if (couchbase) {
+            def result = couchbase.queryN1QL(simple("DELETE FROM $COUCHBASE_BUCKET_NAME"))
```

Comment:
this can have side effects if other data is colocated with the session data in Couchbase, which happens a lot... (I know this is for integration test, but still...) maybe the DELETE query can have a restricting WHERE clause?